File: back/app/services/article_blocks.py
"""아티클 plain text / JSON → 모달용 블록 스키마 (프론트 articleBlocks.js와 동일 의도)."""

from __future__ import annotations

import json
import re
from typing import Any


def _normalize_blocks(blocks: list) -> list[dict[str, Any]]:
    out: list[dict[str, Any]] = []
    for b in blocks or []:
        if not isinstance(b, dict):
            continue
        text = str(b.get("text") or "").strip()
        if not text:
            continue
        item: dict[str, Any] = {
            "type": str(b.get("type") or "paragraph"),
            "text": text,
        }
        if b.get("attribution"):
            item["attribution"] = str(b["attribution"])
        out.append(item)
    return out


def _structured_payload_to_blocks(data: dict) -> list[dict[str, Any]]:
    out: list[dict[str, Any]] = []
    lead = str(data.get("lead") or "").strip()
    if lead:
        out.append({"type": "lead", "text": lead})
    for sec in data.get("sections") or []:
        if not isinstance(sec, dict):
            continue
        heading = str(sec.get("heading") or sec.get("title") or "").strip()
        if heading:
            out.append({"type": "subheader", "text": heading})
        paras = sec.get("paragraphs") or ([sec["text"]] if sec.get("text") else [])
        for p in paras:
            text = str(p or "").strip()
            if text:
                out.append({"type": "paragraph", "text": text})
    quote = data.get("quote")
    if isinstance(quote, dict) and quote.get("text"):
        out.append(
            {
                "type": "quote",
                "text": str(quote["text"]).strip(),
                "attribution": str(quote.get("attribution") or "— 방문 후기"),
            }
        )
    tip = str(data.get("visit_tip") or "").strip()
    if tip:
        out.append({"type": "subheader", "text": "방문 전 알아두면 좋은 것"})
        out.append({"type": "paragraph", "text": tip})
    return out


def try_parse_json_blocks(raw: str) -> list[dict[str, Any]] | None:
    t = str(raw or "").strip()
    if not t.startswith("{") and not t.startswith("["):
        return None
    try:
        data = json.loads(t)
    except json.JSONDecodeError:
        return None
    if isinstance(data, list):
        return _normalize_blocks(data)
    if isinstance(data, dict) and isinstance(data.get("blocks"), list):
        return _normalize_blocks(data["blocks"])
    if isinstance(data, dict) and (data.get("lead") or data.get("sections")):
        return _structured_payload_to_blocks(data)
    return None


_SECTION_RULES: list[tuple[str, re.Pattern[str]]] = [
    ("추천 메뉴", re.compile(r"메뉴|맛집|요리|음식|라자냐|파스타|스테이크|시그니처", re.I)),
    ("분위기 & 뷰", re.compile(r"분위기|뷰|야경|전망|인테리어|공간|감성|데이트", re.I)),
    ("영업 시간", re.compile(r"영업\s*시간|운영\s*시간|브레이크|휴무|\d{1,2}\s*:\s*\d{2}", re.I)),
    ("주차 & 오시는 길", re.compile(r"주차|대중교통|지하철|버스|찾아가", re.I)),
    ("방문 팁", re.compile(r"추천|예약|대기|방문|팁|알아두", re.I)),
]


def _split_sentences(text: str) -> list[str]:
    normalized = re.sub(r"\s+", " ", str(text or "").replace("\n", " ")).strip()
    if not normalized:
        return []
    parts = re.split(r"(?<=[.!?。])\s+(?=[가-힣A-Z「])", normalized)
    return [p.strip() for p in parts if p.strip()] or [normalized]


def _classify_sentence(sentence: str) -> str:
    for header, pattern in _SECTION_RULES:
        if pattern.search(sentence):
            return header
    return "이곳의 이야기"


def content_to_blocks(content: str) -> list[dict[str, Any]]:
    parsed = try_parse_json_blocks(content)
    if parsed:
        return parsed

    text = str(content or "").strip()
    if not text:
        return []

    if "\n\n" in text:
        parts = [p.strip() for p in re.split(r"\n\n+", text) if p.strip()]
        if len(parts) >= 2:
            blocks: list[dict[str, Any]] = [{"type": "lead", "text": parts[0]}]
            for i, part in enumerate(parts[1:], start=1):
                lines = part.split("\n", 1)
                first = lines[0].strip()
                is_header = (
                    len(first) <= 28
                    and not re.search(r"[.!?。]$", first)
                    and (len(lines) > 1 or (i == 1 and len(parts) > 2))
                )
                if is_header and len(lines) > 1:
                    blocks.append({"type": "subheader", "text": first})
                    body = lines[1].strip()
                    if body:
                        blocks.append({"type": "paragraph", "text": body})
                else:
                    blocks.append({"type": "paragraph", "text": part})
            return blocks

    sentences = _split_sentences(text)
    lead_count = min(2, max(1, len(sentences) // 6 or 1))
    lead = " ".join(sentences[:lead_count])
    body = sentences[lead_count:]

    blocks = []
    if lead:
        blocks.append({"type": "lead", "text": lead})

    groups: dict[str, list[str]] = {}
    order: list[str] = []
    for s in body:
        key = _classify_sentence(s)
        if key not in groups:
            groups[key] = []
            order.append(key)
        groups[key].append(s)

    for header in order:
        sents = groups[header]
        if not sents:
            continue
        blocks.append({"type": "subheader", "text": header})
        chunk = " ".join(sents)
        blocks.append({"type": "paragraph", "text": chunk})

    if len(blocks) <= 1 and text:
        return [
            {"type": "lead", "text": text[:280]},
            *(
                [{"type": "subheader", "text": "상세 정보"}, {"type": "paragraph", "text": text[280:]}]
                if len(text) > 280
                else []
            ),
        ]
    return blocks


def serialize_structured_article(
    title: str,
    lead: str,
    sections: list[dict[str, Any]],
    quote: dict[str, str] | None = None,
    visit_tip: str = "",
) -> str:
    payload: dict[str, Any] = {
        "version": 2,
        "title": title,
        "lead": lead,
        "sections": sections,
    }
    if quote:
        payload["quote"] = quote
    if visit_tip:
        payload["visit_tip"] = visit_tip
    return json.dumps(payload, ensure_ascii=False)

File: front/src/utils/articleBlocks.js
/**
 * API 아티클(plain text / JSON) → RegionModal ArticleBody 블록 스키마
 * types: lead | subheader | paragraph | quote
 */

const SECTION_RULES = [
  { header: '추천 메뉴', test: s => /메뉴|맛집|요리|음식|라자냐|파스타|스테이크|코스|시그니처|대표/.test(s) },
  { header: '분위기 & 뷰', test: s => /분위기|뷰|야경|전망|인테리어|공간|감성|로맨틱|데이트/.test(s) },
  { header: '영업 시간', test: s => /영업\s*시간|운영\s*시간|브레이크|휴무|라스트\s*오더|\d{1,2}\s*:\s*\d{2}/.test(s) },
  { header: '주차 & 오시는 길', test: s => /주차|대중교통|지하철|버스|역\s*근처|찾아가|주소/.test(s) },
  { header: '방문 팁', test: s => /추천|예약|대기|방문|팁|알아두|주말|평일/.test(s) },
];

function tryParseJsonBlocks(raw) {
  const t = String(raw || '').trim();
  if (!t.startsWith('{') && !t.startsWith('[')) {
    return null;
  }
  try {
    const data = JSON.parse(t);
    if (Array.isArray(data)) {
      return normalizeBlocks(data);
    }
    if (Array.isArray(data?.blocks)) {
      return normalizeBlocks(data.blocks);
    }
    if (data?.lead || data?.sections) {
      return structuredPayloadToBlocks(data);
    }
  } catch {
    return null;
  }
  return null;
}

function normalizeBlocks(blocks) {
  return blocks
    .filter(b => b && b.text && String(b.text).trim())
    .map(b => ({
      type: b.type || 'paragraph',
      text: String(b.text).trim(),
      attribution: b.attribution || undefined,
    }));
}

function structuredPayloadToBlocks(data) {
  const out = [];
  if (data.lead) {
    out.push({ type: 'lead', text: String(data.lead).trim() });
  }
  for (const sec of data.sections || []) {
    const heading = String(sec.heading || sec.title || '').trim();
    if (heading) {
      out.push({ type: 'subheader', text: heading });
    }
    const paras = sec.paragraphs || (sec.text ? [sec.text] : []);
    for (const p of paras) {
      const text = String(p || '').trim();
      if (text) {
        out.push({ type: 'paragraph', text });
      }
    }
  }
  if (data.quote?.text) {
    out.push({
      type: 'quote',
      text: String(data.quote.text).trim(),
      attribution: data.quote.attribution || '— 방문 후기',
    });
  }
  if (data.visit_tip) {
    out.push({ type: 'subheader', text: '방문 전 알아두면 좋은 것' });
    out.push({ type: 'paragraph', text: String(data.visit_tip).trim() });
  }
  return out;
}

function splitSentences(text) {
  const normalized = String(text || '')
    .replace(/\r\n/g, '\n')
    .replace(/\n+/g, ' ')
    .replace(/\s+/g, ' ')
    .trim();
  if (!normalized) {
    return [];
  }

  const sentences = [];
  let buf = '';
  for (let i = 0; i < normalized.length; i += 1) {
    const ch = normalized[i];
    buf += ch;
    if (ch === '.' || ch === '!' || ch === '?' || ch === '。') {
      const tail = normalized.slice(i + 1, i + 4);
      if (/^\s*\d/.test(tail)) {
        continue;
      }
      const piece = buf.trim();
      if (piece) {
        sentences.push(piece);
      }
      buf = '';
    }
  }
  if (buf.trim()) {
    sentences.push(buf.trim());
  }
  return sentences.length > 0 ? sentences : [normalized];
}

function extractQuote(sentences) {
  const rest = [];
  let quote = null;
  for (const s of sentences) {
    const m = s.match(/[""]([^""]+)[""]/) || s.match(/['']([^'']+)['']/);
    if (m && m[1].length > 12 && m[1].length < 120) {
      quote = { type: 'quote', text: `"${m[1].trim()}"`, attribution: '— 방문 후기' };
      const cleaned = s.replace(m[0], '').trim();
      if (cleaned) {
        rest.push(cleaned);
      }
    } else {
      rest.push(s);
    }
  }
  return { sentences: rest, quote };
}

function classifySentence(sentence) {
  for (const rule of SECTION_RULES) {
    if (rule.test(sentence)) {
      return rule.header;
    }
  }
  return '이곳의 이야기';
}

function groupSentencesBySection(sentences) {
  const groups = new Map();
  const order = [];

  for (const s of sentences) {
    const key = classifySentence(s);
    if (!groups.has(key)) {
      groups.set(key, []);
      order.push(key);
    }
    groups.get(key).push(s);
  }

  return order.map(header => ({ header, sentences: groups.get(header) }));
}

function splitParagraphChunks(sentences, maxChars = 320) {
  const chunks = [];
  let buf = '';

  for (const s of sentences) {
    const next = buf ? `${buf} ${s}` : s;
    if (next.length > maxChars && buf) {
      chunks.push(buf);
      buf = s;
    } else {
      buf = next;
    }
  }
  if (buf) {
    chunks.push(buf);
  }
  return chunks;
}

/** plain text / JSON → ArticleBody blocks */
export function parseArticleContentToBlocks(content, region = {}) {
  const fromJson = tryParseJsonBlocks(content);
  if (fromJson?.length) {
    return fromJson;
  }

  const text = String(content || '').trim();
  if (!text) {
    return [];
  }

  if (text.includes('\n\n')) {
    const parts = text.split(/\n\n+/).map(p => p.trim()).filter(Boolean);
    if (parts.length >= 2) {
      const blocks = [{ type: 'lead', text: parts[0] }];
      for (let i = 1; i < parts.length; i += 1) {
        const part = parts[i];
        const firstLine = part.split('\n')[0].trim();
        const isShortHeader =
          firstLine.length <= 28 &&
          !/[.!?。]$/.test(firstLine) &&
          part.includes('\n');
        if (isShortHeader) {
          blocks.push({ type: 'subheader', text: firstLine });
          const body = part.slice(firstLine.length).trim();
          if (body) {
            blocks.push({ type: 'paragraph', text: body });
          }
        } else if (
          firstLine.length <= 24 &&
          !/[.!?。]$/.test(firstLine) &&
          i === 1 &&
          parts.length > 2
        ) {
          blocks.push({ type: 'subheader', text: firstLine });
          blocks.push({
            type: 'paragraph',
            text: part.slice(firstLine.length).trim() || part,
          });
        } else {
          blocks.push({ type: 'paragraph', text: part });
        }
      }
      return blocks;
    }
  }

  let sentences = splitSentences(text);
  const { sentences: withoutQuote, quote } = extractQuote(sentences);
  sentences = withoutQuote;

  const leadCount = Math.min(2, Math.max(1, Math.floor(sentences.length * 0.15)));
  const lead = sentences.slice(0, leadCount).join(' ');
  const bodySentences = sentences.slice(leadCount);

  const blocks = [];
  if (lead) {
    blocks.push({ type: 'lead', text: lead });
  }

  const groups = groupSentencesBySection(bodySentences);
  for (const { header, sentences: groupSents } of groups) {
    if (!groupSents.length) {
      continue;
    }
    blocks.push({ type: 'subheader', text: header });
    for (const para of splitParagraphChunks(groupSents)) {
      blocks.push({ type: 'paragraph', text: para });
    }
  }

  if (quote) {
    blocks.push(quote);
  }

  if (blocks.length <= 1 && text) {
    return [
      { type: 'lead', text: text.slice(0, Math.min(280, text.length)) },
      ...(text.length > 280
        ? [{ type: 'subheader', text: '상세 정보' }, { type: 'paragraph', text: text.slice(280) }]
        : []),
    ];
  }

  return blocks;
}

export function buildArticleDisplayData(article, region) {
  const name = region?.name || '이 장소';
  const regionName = region?.region || region?.province || '이 지역';
  const summary = String(region?.summary || '').trim();

  const fallback = {
    title: `${name}, 로컬이 사랑하는 이유`,
    author: 'LocalVibe 에디터',
    date: new Date().toLocaleDateString('ko-KR', {
      year: 'numeric',
      month: 'long',
      day: 'numeric',
    }),
    body: [
      {
        type: 'lead',
        text: `${regionName}에서 진짜 로컬을 만나고 싶다면, 관광지도를 잠시 접어두자. ${name}은 그런 곳이다.`,
      },
      { type: 'subheader', text: '지역이 품은 이야기' },
      {
        type: 'paragraph',
        text:
          summary ||
          `${name}은 입소문을 타며 여행객에게도 알려지기 시작한 공간이다.`,
      },
      { type: 'subheader', text: '공간이 주는 감각' },
      {
        type: 'paragraph',
        text: `${name}에 들어서면 서두르지 않아도 된다는 느낌이 든다. 테이블, 음악, 창밖 풍경이 자연스럽게 어우러진다.`,
      },
      {
        type: 'quote',
        text: `"처음 왔을 때 그냥 지나칠 뻔했어요. 이제는 ${regionName} 오면 꼭 들르는 곳이 됐어요."`,
        attribution: '— 방문객 후기',
      },
      { type: 'subheader', text: '방문 전 알아두면 좋은 것' },
      {
        type: 'paragraph',
        text: `주말 오후는 붐빌 수 있다. 여유롭게 즐기려면 평일 오전·저녁을 추천한다.`,
      },
    ],
  };

  const hasReal =
    article &&
    (article.title || article.content) &&
    !String(article.content || '').includes('오류') &&
    !String(article.content || '').includes('불러오지 못') &&
    String(article.content || '').length > 30;

  if (!hasReal) {
    return fallback;
  }

  const blocksFromApi =
    Array.isArray(article.blocks) && article.blocks.length > 0
      ? normalizeBlocks(article.blocks)
      : parseArticleContentToBlocks(article.content, region);

  const body =
    blocksFromApi.length > 0 ? blocksFromApi : fallback.body;

  return {
    title: String(article.title || '').trim() || fallback.title,
    author: 'LocalVibe AI',
    date: '',
    body,
  };
}
